### PR TITLE
[dataclasses plugin] Support kw_only=True

### LINKED ISSUE #10865: Don't warn about non keyword-only attributes after keyword-only attributes in dataclasses
**Bug Report**

If all fields in a dataclass are keyword only then mypy should not raise errors regarding the order of attributes with default values.

Python 3.10.0 introduces a new dataclass parameter [`kw_only`](https://docs.python.org/3.10/library/dataclasses.html#dataclasses.dataclass).

When this parameter is set to true then all fields in the dataclass will be marked as keyword only. If all fields in a dataclass are keyword only then checking the order of attributes with default values is not relevant.

Prior art:

[attrs plugin] Support kw_only=True #6107
Allow attrs kw_only arguments at any position #8803

bpo-43532: Add keyword-only fields to dataclasses. https://github.com/python/cpython/pull/25608
Sync typeshed #10537 (introduced the `kw_only` dataclass parameter to mypy)

**To Reproduce**

1. Ensure you are using python version `3.10.0b4` (although should be reproducible with other 3.10 beta versions)
2. Ensure you are using a dev build of mypy off of master
3. Create a file `test_dataclass.py` which contains the following
```python
from dataclasses import dataclass


@dataclass(kw_only=True)
class Example:
    name: str = 'hi'
    email: str
```
4. Run mypy and observe that it fails
```bash
$ mypy test_dataclass.py 
test_dataclass.py:7: error: Attributes without a default cannot follow attributes with one
Found 1 error in 1 file (checked 1 source file)
```

**Expected Behavior**

This should not cause a mypy error. Checking the ordering of attributes with default parameters is not necessary when `kw_only=True` is specified.

**Your Environment**

- Mypy version used: 0.920+dev.97b3b90356893e2a9f6947ebe52b41cd54ab1cb7
- Mypy command-line flags: mypy test_dataclass.py
- Mypy configuration options from `mypy.ini` (and other config files): n/a; default
- Python version used: 3.10.0b4
- Operating system and version: OS X 10.15.5

### PR DESCRIPTION
**Pull request by tgallant**
### Description

Fixes #10865

If all fields in a dataclass are keyword only then mypy should not raise errors regarding the order of attributes with default values. See the linked issue for a more detailed description.

This PR makes the relevant changes to handle that.

## Test Plan

I've added tests to `check-dataclasses.test` to cover the new changes.

### CODE REVIEW

File: test-data/unit/lib-stub/dataclasses.pyi [hunk 1]
```diff
@@ -1,30 +1,86 @@
...
 from typing import Any, Callable, Generic, Mapping, Optional, TypeVar, overload, Type
 
 _T = TypeVar('_T')
 
 class InitVar(Generic[_T]):
     ...
 
+if sys.version_info >= (3, 10):
+    class _KW_ONLY_TYPE: ...
+    KW_ONLY = _KW_ONLY_TYPE
```
Comment (tgallant) [inline test-data/unit/lib-stub/dataclasses.pyi:11]: in the stub from typeshed this line is

```python
    KW_ONLY: _KW_ONLY_TYPE
```

Without making this change the following will fail

```python
[case testDataclassesOrderingKwOnlyWithSentinel]
# flags: --python-version 3.10
from dataclasses import dataclass, KW_ONLY

@dataclass
class Application:
    _: KW_ONLY
    name: str = 'Unnamed'
    rating: int
```

```
error: Variable "dataclasses.KW_ONLY" is not valid as a type
```

Any feedback on this? Does this change need to be made in typeshed or am I specifying the type wrong in the test case?

Comment (JelleZijlstra) [reply test-data/unit/lib-stub/dataclasses.pyi:11]: I think typeshed has to change. I believe that test case is how the object is supposed to be used, so typeshed should accommodate it.

**Comment by cdce8p** [comment]
Came across an interesting use case of `KW_ONLY` with inheritance, that should probably be supported as well (if not already done).
```py
from dataclasses import dataclass, KW_ONLY

@dataclass
class Base:
    x: str
    _: KW_ONLY
    y: int = 0
    w: int = 1

@dataclass
class D(Base):
    z: str
    a: str = "a"
```
```py
>>> D("Hello", "World")
D(x='Hello', y=0, w=1, z='World', a='a')
```
That works because of keyword reordering.
https://docs.python.org/3.10/library/dataclasses.html#re-ordering-of-keyword-only-parameters-in-init

**Comment by hauntsaninja** [comment]
Thanks for working on this! Could we also test that `__init__` has the appropriate signature (e.g. by instantiating the classes)

**Comment by tgallant** [comment]
Thanks for raising this case @cdce8p. I've added this example as a test case and updated the code to handle it. @hauntsaninja I also made sure to check the initialization in the test cases.

I was looking through the docs again and spotted this:

> In a single dataclass, it is an error to specify more than one field whose type is KW_ONLY.

https://docs.python.org/3.10/library/dataclasses.html#dataclasses.KW_ONLY

It seemed like this would be a good thing to have a mypy check for. I added a check for this case in commit https://github.com/python/mypy/pull/10867/commits/75ef138142b6a7da173730afa28ab49f2e4e2374

File: test-data/unit/lib-stub/dataclasses.pyi [hunk 2]
```diff
@@ -5,26 +5,29 @@ _T = TypeVar('_T')
 class InitVar(Generic[_T]):
     ...
 
+class _KW_ONLY_TYPE: ...
+KW_ONLY = _KW_ONLY_TYPE
```
Comment (cdce8p) [inline test-data/unit/lib-stub/dataclasses.pyi:9]: Not entirely sure, but https://github.com/python/typeshed/pull/5826 changed how `KW_ONLY` is defined.
After the typeshed changes have been synced, I think this needs to be updated.

**Comment by hauntsaninja** [comment]
python/typeshed#5826 is now on mypy master, so if you merge you'll be able to pull that in.

It would be good to add tests of class instantiation that produce errors. Specifically, test cases should prove that the arguments are in fact keyword only and complain when passed positionally and that parameter reordering happens and we get type errors (i.e. make sure positional parameters have types that are distinct from each other and from kw only parameters)

**Comment by tgallant** [comment]
i rebased this branch with master and updated the `KW_ONLY` usage according to the typeshed change. i will follow up with some more test cases

**Comment by tgallant** [comment]
thanks for the helpful suggestions @hauntsaninja and @cdce8p. i've added test cases covering positional vs keyword arguments, reordering, and type errors. let me know if you think there are other cases that should be covered in this PR.

**Review by hauntsaninja — APPROVED** [review]
Thanks, this is great! :-)

**Comment by cdce8p** [comment]
Thanks @tgallant 🎉

**Pull request merged**